kernel/src/table_properties.rs

```diff
@@ -0,0 +1,374 @@
+    #[serde(flatten)]
+    pub unknown_properties: HashMap<String, String>,
+}
+
+impl TableProperties {
+    pub(crate) fn new<'a, I>(items: I) -> DeltaResult<Self>
+    where
+        I: Iterator<Item = (&'a String, &'a String)> + 'a,
+    {
+        let deserializer = MapDeserializer::<_, de::value::Error>::new(
+            items.map(|(k, v)| (k.as_str(), v.as_str())),
+        );
+        TableProperties::deserialize(deserializer).map_err(Error::invalid_table_properties)
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, Clone, Deserialize)]
+#[serde(try_from = "String")]
+pub enum DataSkippingNumIndexedCols {
+    AllColumns,
+    NumColumns(u64),
+}
+
+impl TryFrom<String> for DataSkippingNumIndexedCols {
+    type Error = Error;
+
+    fn try_from(value: String) -> Result<Self, Self::Error> {
+        let num: i64 = value.parse().map_err(|_| {
+            Error::invalid_table_properties(
+                "couldn't parse DataSkippingNumIndexedCols to an integer",
+            )
+        })?;
+        match num {
+            -1 => Ok(DataSkippingNumIndexedCols::AllColumns),
+            x => Ok(DataSkippingNumIndexedCols::NumColumns(
+                x.try_into().map_err(|_| {
+                    Error::invalid_table_properties(
+                        "couldn't parse DataSkippingNumIndexedCols to positive integer",
+                    )
+                })?,
+            )),
+        }
+    }
+}
+
+/// The isolation level applied during transaction
+#[derive(Deserialize, Debug, Copy, Clone, PartialEq, Eq)]
+#[serde(rename_all = "camelCase")]
+pub enum IsolationLevel {
+    /// The strongest isolation level. It ensures that committed write operations
+    /// and all reads are Serializable. Operations are allowed as long as there
+    /// exists a serial sequence of executing them one-at-a-time that generates
+    /// the same outcome as that seen in the table. For the write operations,
+    /// the serial sequence is exactly the same as that seen in the table’s history.
+    Serializable,
+
+    /// A weaker isolation level than Serializable. It ensures only that the write
+    /// operations (that is, not reads) are serializable. However, this is still stronger
+    /// than Snapshot isolation. WriteSerializable is the default isolation level because
+    /// it provides great balance of data consistency and availability for most common operations.
+    WriteSerializable,
+
+    /// SnapshotIsolation is a guarantee that all reads made in a transaction will see a consistent
+    /// snapshot of the database (in practice it reads the last committed values that existed at the
+    /// time it started), and the transaction itself will successfully commit only if no updates
+    /// it has made conflict with any concurrent updates made since that snapshot.
+    SnapshotIsolation,
+}
+
+// Delta-Spark default isolation level is Serializable
+impl Default for IsolationLevel {
+    fn default() -> Self {
+        Self::Serializable
+    }
+}
```

Comment:
Does `#[derive(Default)]` not choose the first enum variant automatically?

(again below)

Reply:
hm doesn't look like it but discovered via clippy we can annotate the default variant with `#[default]`!

Reply:
noice!